### ISSUE TITLE
OpenJDK java/lang/Thread/virtual/stress/PingPong timeout 

### ISSUE THREAD

**Issue by pshipton** [eclipse-openj9/openj9#17255]
https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/40
jdk_lang_j9_0
`-Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage -Xjit -Xgcpolicy:balanced -Xnocompressedrefs`
java/lang/Thread/virtual/stress/PingPong.java

```
20:59:29  Timeout information:
20:59:29  Running jcmd on process 2680997
20:59:29  Dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16821212678666/jdk_lang_j9_0/work/scratch/0/core.20230422.005853.2680997.0001.dmp
20:59:29  Dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16821212678666/jdk_lang_j9_0/work/scratch/0/javacore.20230422.005925.2680997.0002.txt
20:59:29  Running jstack on process 2680997
20:59:29  2023-04-22T00:59:26.449132873
20:59:29  Virtual machine: 2680997 JVM information:
20:59:29  JRE 20 Linux s390x-64-Bit 20230421_43 (JIT enabled, AOT enabled)
20:59:29  OpenJ9   - bf214ec2114
20:59:29  OMR      - 97cf2a6e860
20:59:29  JCL      - 38d10ea9b95 based on jdk-20+36
20:59:29  
20:59:29  "main" prio=5 Id=2 WAITING
20:59:29  	at java.base@20-internal/java.lang.Object.waitImpl(Native Method)
20:59:29  	at java.base@20-internal/java.lang.Object.wait(Object.java:255)
20:59:29  	at java.base@20-internal/java.lang.Object.wait(Object.java:221)
20:59:29  	at java.base@20-internal/java.lang.Thread.join(Thread.java:2147)
20:59:29  	at java.base@20-internal/java.lang.Thread.join(Thread.java:2223)
20:59:29  	at app//com.sun.javatest.regtest.agent.MainWrapper.main(MainWrapper.java:72)
20:59:29  
20:59:29  "JIT Compilation Thread-000" prio=10 Id=4 RUNNABLE
20:59:29  
20:59:29  "JIT Compilation Thread-001 Suspended" prio=10 Id=5 RUNNABLE
20:59:29  
20:59:29  "JIT Compilation Thread-002 Suspended" prio=10 Id=6 RUNNABLE
20:59:29  
20:59:29  "JIT Compilation Thread-003 Suspended" prio=10 Id=7 RUNNABLE
20:59:29  
20:59:29  "JIT Compilation Thread-004 Suspended" prio=10 Id=8 RUNNABLE
20:59:29  
20:59:29  "JIT Compilation Thread-005 Suspended" prio=10 Id=9 RUNNABLE
20:59:29  
20:59:29  "JIT Compilation Thread-006 Suspended" prio=10 Id=10 RUNNABLE
20:59:29  
20:59:29  "JIT Diagnostic Compilation Thread-007 Suspended" prio=10 Id=11 RUNNABLE
20:59:29  
20:59:29  "JIT-SamplerThread" prio=10 Id=12 TIMED_WAITING
20:59:29  
20:59:29  "IProfiler" prio=5 Id=13 RUNNABLE
20:59:29  
20:59:29  "Common-Cleaner" prio=8 Id=3 TIMED_WAITING
20:59:29  	at java.base@20-internal/java.lang.Object.waitImpl(Native Method)
20:59:29  	at java.base@20-internal/java.lang.Object.wait(Object.java:255)
20:59:29  	at java.base@20-internal/java.lang.Object.wait(Object.java:221)
20:59:29  	at java.base@20-internal/java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:137)
20:59:29  	at java.base@20-internal/jdk.internal.ref.CleanerImpl.run(CleanerImpl.java:140)
20:59:29  	at java.base@20-internal/java.lang.Thread.run(Thread.java:1639)
20:59:29  	at java.base@20-internal/jdk.internal.misc.InnocuousThread.run(InnocuousThread.java:186)
20:59:29  
20:59:29  "Dedicated GC Main" prio=5 Id=14 RUNNABLE
20:59:29  
20:59:29  "GC Worker" prio=5 Id=15 RUNNABLE
20:59:29  
20:59:29  "GC Worker" prio=5 Id=16 RUNNABLE
20:59:29  
20:59:29  "GC Worker" prio=5 Id=17 RUNNABLE
20:59:29  
20:59:29  "Attach API wait loop" prio=10 Id=20 RUNNABLE
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.IPC.waitSemaphore(Native Method)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.CommonDirectory.waitSemaphore(CommonDirectory.java:264)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.WaitLoop.waitForNotification(WaitLoop.java:66)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.WaitLoop.run(WaitLoop.java:157)
20:59:29  
20:59:29  "MainThread" prio=5 Id=22 TIMED_WAITING
20:59:29  	at java.base@20-internal/jdk.internal.misc.Unsafe.park(Native Method)
20:59:29  	at java.base@20-internal/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:269)
20:59:29  	at java.base@20-internal/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:756)
20:59:29  	at java.base@20-internal/java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1126)
20:59:29  	at java.base@20-internal/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:276)
20:59:29  	at java.base@20-internal/java.lang.VirtualThread.joinNanos(VirtualThread.java:816)
20:59:29  	at java.base@20-internal/java.lang.Thread.join(Thread.java:2261)
20:59:29  	at app//PingPong.main(PingPong.java:95)
20:59:29  	at java.base@20-internal/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
20:59:29  	at java.base@20-internal/java.lang.reflect.Method.invoke(Method.java:578)
20:59:29  	at app//com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:125)
20:59:29  	at java.base@20-internal/java.lang.Thread.run(Thread.java:1639)
20:59:29  
20:59:29  "Finalizer thread" prio=5 Id=29 RUNNABLE
20:59:29  
20:59:29  "file lock watchdog" prio=10 Id=31 TIMED_WAITING
20:59:29  	at java.base@20-internal/java.lang.Object.waitImpl(Native Method)
20:59:29  	at java.base@20-internal/java.lang.Object.wait(Object.java:255)
20:59:29  	at java.base@20-internal/java.lang.Object.wait(Object.java:221)
20:59:29  	at java.base@20-internal/java.util.TimerThread.mainLoop(Timer.java:599)
20:59:29  	at java.base@20-internal/java.util.TimerThread.run(Timer.java:542)
20:59:29  
20:59:29  "Attachment portNumber: 38697" prio=10 Id=33 RUNNABLE
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.DiagnosticUtils.dumpAllThreadsImpl(Native Method)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.DiagnosticUtils.getThreadInfo(DiagnosticUtils.java:245)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.DiagnosticUtils.executeDiagnosticCommand(DiagnosticUtils.java:181)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.Attachment.doCommand(Attachment.java:248)
20:59:29  	at java.base@20-internal/openj9.internal.tools.attach.target.Attachment.run(Attachment.java:159)
20:59:29  
20:59:29  
20:59:29  --- Timeout information end.

20:59:29  STDOUT:
...
20:59:29  252617 252617
20:59:29  252617 252617
20:59:29  252617 252617
20:59:29  Timeout refired 960 times
20:59:29  252617 252617
20:59:29  252617 252617
20:59:29  252617 252617
...
```

**Comment by pshipton** [comment]
@fengxue-IS @babsingh @tajila fyi

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_OpenJDK20/3/

**Comment by tajila** [comment]
There is a crash in https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/41/consoleFull

```
02:59:12          -J-Dtest.class.path.prefix=/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/classes/1/java/lang/String/concat/ImplicitStringConcatShapes.d:/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/String/concat \
02:59:12          -d /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/classes/1/java/lang/String/concat/ImplicitStringConcatShapes.d \
02:59:12          -sourcepath /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/String/concat \
02:59:12          -classpath /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/String/concat:/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/classes/1/java/lang/String/concat/ImplicitStringConcatShapes.d /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/openjdk/openjdk-jdk/test/jdk/java/lang/String/concat/ImplicitStringConcatShapes.java
02:59:12  stderr:
02:59:12  Unhandled exception
02:59:12  Type=Segmentation error vmState=0x00000000
02:59:12  J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=000000a9 Signal_Code=00000001
02:59:12  Handler1=000003FFA73C48F8 Handler2=000003FFA72B1998 InaccessibleAddress=000000000002F000
02:59:12  gpr0=000003FFA060E200 gpr1=000003FF7C850DF0 gpr2=000003FF7CA50A18 gpr3=000003FF001219EF
02:59:12  gpr4=000000000000000D gpr5=000003FFA10CA1F0 gpr6=000003FF001219EF gpr7=000003FF7C850DF0
02:59:12  gpr8=0000000000030000 gpr9=000003FF7AE5D090 gpr10=000003FF7C85E0E8 gpr11=000003FF7C80D658
02:59:12  gpr12=000003FF001219EF gpr13=000003FFA0121600 gpr14=000003FF7C97FAC8 gpr15=000003FFA7F74320
02:59:12  psw=000003FF6123AB3C mask=0705100180000000 fpc=0048fe00 bea=000003FF61238AEA
02:59:12  fpr0 488df6b300000000 (f: 0.000000, d: 3.262755e+41)
02:59:12  fpr1 be18f00000000000 (f: 0.000000, d: -1.451554e-09)
02:59:12  fpr2 bddf6b3700000000 (f: 0.000000, d: -1.143010e-10)
02:59:12  fpr3 3dd064a800000000 (f: 0.000000, d: 5.963807e-11)
02:59:12  fpr4 3dd40d4840000000 (f: 1073741824.000000, d: 7.294833e-11)
02:59:12  fpr5 3f31720000000000 (f: 0.000000, d: 2.661943e-04)
02:59:12  fpr6 b17e5dbf00000000 (f: 0.000000, d: -2.749865e-70)
02:59:12  fpr7 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:59:12  fpr8 00000000123565a0 (f: 305489312.000000, d: 1.509318e-315)
02:59:12  fpr9 0000000000555b78 (f: 5593976.000000, d: 2.763791e-317)
02:59:12  fpr10 00000000123568a8 (f: 305490080.000000, d: 1.509322e-315)
02:59:12  fpr11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:59:12  fpr12 0005fa238ee8101b (f: 2397573120.000000, d: 8.312188e-309)
02:59:12  fpr13 000003fef0043ce8 (f: 4026809600.000000, d: 2.170669e-311)
02:59:12  fpr14 0000000000555b80 (f: 5593984.000000, d: 2.763795e-317)
02:59:12  fpr15 000003fef0043da8 (f: 4026809856.000000, d: 2.170669e-311)
02:59:12  
02:59:12  Compiled_method=sun/nio/cs/UTF_8$Decoder.decodeArrayLoop(Ljava/nio/ByteBuffer;Ljava/nio/CharBuffer;)Ljava/nio/charset/CoderResult;
02:59:12  Target=2_90_20230425_44 (Linux 3.10.0-1160.83.1.el7.s390x)
02:59:12  CPU=s390x (4 logical CPUs) (0x1ec1b1000 RAM)
02:59:12  ----------- Stack Backtrace -----------
02:59:12   (0x000003FF6123AB3C [<unknown>+0x0])
02:59:12  ---------------------------------------
02:59:12  JVMDUMP039I Processing dump event "gpf", detail "" at 2023/04/25 02:58:43 - please wait.
02:59:12  JVMDUMP032I JVM requested System dump using '/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/core.20230425.025843.17925.0001.dmp' in response to an event
02:59:12  JVMDUMP010I System dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/core.20230425.025843.17925.0001.dmp
02:59:12  JVMDUMP032I JVM requested Java dump using '/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/javacore.20230425.025843.17925.0002.txt' in response to an event
02:59:12  Unhandled exception
02:59:12  Type=Segmentation error vmState=0x00000000
02:59:12  J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=000000a9 Signal_Code=00000001
02:59:12  Handler1=000003FFA73C48F8 Handler2=000003FFA72B1998 InaccessibleAddress=0000000000000000
02:59:12  gpr0=0000000000000001 gpr1=0000000000000000 gpr2=0000000000000001 gpr3=000003FF04019E00
02:59:12  gpr4=0000000000000000 gpr5=000003FF7C962EC0 gpr6=000003FFA6CFDCD0 gpr7=000003FFA6CFDD89
02:59:12  gpr8=000000000000000A gpr9=000003FF7C962EB8 gpr10=000003FFA5AADD78 gpr11=000003FFA0015430
02:59:12  gpr12=000003FF609C6DE8 gpr13=000003FFA5CBC4E0 gpr14=000003FFA5AC30A0 gpr15=000003FFA6CFDA30
02:59:12  psw=000003FFA5AC30AC mask=0705200180000000 fpc=0008fe00 bea=000003FFA5AB3BFC
02:59:12  fpr0 428e5efe00000000 (f: 0.000000, d: 4.174167e+12)
02:59:12  fpr1 3e63917800000000 (f: 0.000000, d: 3.644868e-08)
02:59:12  fpr2 3e3ab28300000000 (f: 0.000000, d: 6.215952e-09)
02:59:12  fpr3 3c14bc0100000000 (f: 0.000000, d: 2.810034e-19)
02:59:12  fpr4 be386333a0287658 (f: 2687006208.000000, d: -5.678159e-09)
02:59:12  fpr5 3e924b1c00000000 (f: 0.000000, d: 2.725928e-07)
02:59:12  fpr6 3ecccf3400000040 (f: 64.000000, d: 3.434347e-06)
02:59:12  fpr7 3e3a365700000000 (f: 0.000000, d: 6.103019e-09)
02:59:12  fpr8 000003ffa7423568 (f: 2806134016.000000, d: 2.172188e-311)
02:59:12  fpr9 0000000000555b78 (f: 5593976.000000, d: 2.763791e-317)
02:59:12  fpr10 000003ffa0015430 (f: 2684441600.000000, d: 2.172128e-311)
02:59:12  fpr11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:59:12  fpr12 0005fa238ee8101b (f: 2397573120.000000, d: 8.312188e-309)
02:59:12  fpr13 000003fef0043ce8 (f: 4026809600.000000, d: 2.170669e-311)
02:59:12  fpr14 0000000000555b80 (f: 5593984.000000, d: 2.763795e-317)
02:59:12  fpr15 000003fef0043da8 (f: 4026809856.000000, d: 2.170669e-311)
02:59:12  Module=/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/openjdkbinary/j2sdk-image/lib/default/libj9gc_full29.so
02:59:12  Module_base_address=000003FFA5A80000
02:59:12  Target=2_90_20230425_44 (Linux 3.10.0-1160.83.1.el7.s390x)
02:59:12  CPU=s390x (4 logical CPUs) (0x1ec1b1000 RAM)
02:59:12  ----------- Stack Backtrace -----------
02:59:12   (0x000003FFA5AC30AC [libj9gc_full29.so+0x430ac])
02:59:12  comparatorToEqualFn+0x28 (0x000003FFA5CA6790 [libj9gc_full29.so+0x226790])
02:59:12  hashTableFind+0xbe (0x000003FFA5CA74C6 [libj9gc_full29.so+0x2274c6])
02:59:12  _ZN14MM_StringTable6hashAtEmP8J9Object+0x2e (0x000003FFA5AC49C6 [libj9gc_full29.so+0x449c6])
02:59:12  _ZN14MM_StringTable10hashAtUTF8EmPhmj+0x32 (0x000003FFA5AC4A1A [libj9gc_full29.so+0x44a1a])
02:59:12  j9gc_createJavaLangString+0xf7c (0x000003FFA5AC608C [libj9gc_full29.so+0x4608c])
02:59:12  resolveStringRef+0x10a (0x000003FFA7403F32 [libj9vm29.so+0x83f32])
02:59:12  _ZN26VM_BytecodeInterpreterFull3runEP10J9VMThread+0x2fa54 (0x000003FFA748FEF4 [libj9vm29.so+0x10fef4])
02:59:12  bytecodeLoopFull+0xee (0x000003FFA746048E [libj9vm29.so+0xe048e])
02:59:12  c_cInterpreter+0x64 (0x000003FFA752CF04 [libj9vm29.so+0x1acf04])
02:59:12  ---------------------------------------
02:59:12  JVMDUMP039I Processing dump event "gpf", detail "" at 2023/04/25 02:58:55 - please wait.
02:59:12  JVMDUMP032I JVM requested System dump using '/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/core.20230425.025855.17925.0003.dmp' in response to an event
02:59:12  JVMDUMP010I Java dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/javacore.20230425.025843.17925.0002.txt
02:59:12  JVMDUMP010I System dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/core.20230425.025855.17925.0003.dmp
02:59:12  JVMDUMP032I JVM requested Java dump using '/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/javacore.20230425.025855.17925.0005.txt' in response to an event
02:59:12  JVMDUMP032I JVM requested Snap dump using '/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/Snap.20230425.025843.17925.0004.trc' in response to an event
02:59:12  JVMDUMP010I Java dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/javacore.20230425.025855.17925.0005.txt
02:59:12  JVMDUMP010I Snap dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/Snap.20230425.025843.17925.0004.trc
02:59:12  JVMDUMP032I JVM requested JIT dump using '/home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/jitdump.20230425.025843.17925.0007.dmp' in response to an event
02:59:12  JVMDUMP051I JIT dump occurred in 'main' thread 0x000003FFA0121600
02:59:12  JVMDUMP053I JIT dump is recompiling sun/nio/cs/UTF_8$Decoder.decodeArrayLoop(Ljava/nio/ByteBuffer;Ljava/nio/CharBuffer;)Ljava/nio/charset/CoderResult;
02:59:12  JVMDUMP053I JIT dump is recompiling com/sun/tools/javac/main/JavaCompiler.readSource(Ljavax/tools/JavaFileObject;)Ljava/lang/CharSequence;
02:59:12  JVMDUMP052I JIT dump recursive crash occurred on diagnostic thread
02:59:12  JVMDUMP053I JIT dump is recompiling com/sun/tools/javac/main/JavaCompiler.compile(Ljava/util/Collection;Ljava/util/Collection;Ljava/lang/Iterable;Ljava/util/Collection;)V
02:59:12  JVMDUMP052I JIT dump recursive crash occurred on diagnostic thread
02:59:12  JVMDUMP053I JIT dump is recompiling com/sun/javatest/regtest/agent/RegressionCompileCommand.run([Ljava/lang/String;Ljava/io/PrintWriter;Ljava/io/PrintWriter;)Lcom/sun/javatest/regtest/agent/AStatus;
02:59:12  JVMDUMP010I JIT dump written to /home/jenkins/workspace/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/aqa-tests/TKG/output_16824043581126/jdk_lang_j9_0/work/scratch/0/jitdump.20230425.025843.17925.0007.dmp
02:59:12  JVMDUMP013I Processed dump event "gpf", detail "".
02:59:12  
02:59:12  TEST RESULT: Error. Agent communication error: java.io.EOFException; check console log for any additional details
02:59:12  --------------------------------------------------
03:16:31  TEST: java/lang/Thread/virtual/stress/PingPong.java#sq
```

**Comment by tajila** [comment]
Ignore the previous post, its unrelated.

**Comment by pshipton** [comment]
We have a different PR for that crash, which is https://github.com/eclipse-openj9/openj9/issues/17203

The PingPong test didn't crash, it timed out. The output for the test is below `03:16:31  TEST: java/lang/Thread/virtual/stress/PingPong.java#sq`

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/43

**Comment by tajila** [comment]
Does this only fail with balanced GC? All the links I've looked at are only balanced

**Comment by pshipton** [comment]
Yes probably, we never saw it before. The jdk_lang_j9 target to test balanced was only recently added.

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/45

**Comment by tajila** [comment]
@LinHu2016 Please take a look/ FYI @fengxue-IS

**Comment by LinHu2016** [comment]
Investigating the timeout
1, could not reproduce the timeout issue locally(x86_64_linux), looks like it is platform specific issue( so far all of reported issues are on zlinux).
2,check from core dump file, there were only two virtual threads, both of them were unmounted(not finished), there were neither STW GC nor concurrent GC active, 
3, check test src code, looks like both virtual threads only could be unmounted during  `queue.take()` and should be waked up(mount) by another virtual thread `queue.transfer("xxx")`, expect there would be 10,000,000 mount/umount for each virtual thread, but after 252617 cycles both virtual threads has been unmounted, and never mount back...
4, so far haven't found any balanced GC actions could cause the situation.

**Comment by pshipton** [comment]
I wonder if it's a dup of https://github.com/eclipse-openj9/openj9/issues/16969. This is just fixed so if we see another failure then it's not a dup.

**Comment by pshipton** [comment]
Another timeout
https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_Nightly/49

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk20_j9_sanity.openjdk_s390x_linux_OMR/3/

**Comment by LinHu2016** [comment]
Investigating update:
1, we can reproduce the issue on zlinux machine with option `--enable-preview -Xnocompressedrefs -XX:-JITServerTechPreviewMessage -Xgcpolicy:balanced  PingPong`, it is intermittent issue( fail between 1/10 to 8/10 only on zlinux).
2, no sign the issue was caused by GC scanning(both STW and concurrent), we can reproduce the issue with -Xgc:noConcurrentMarkKO and for all of failure cases there were no concurrent scanning and no GMP or global collections, we also have two failure cases without any GC collection( before first GC collection).
3, add debug print in  JVM_VirtualThreadMountBegin , JVM_VirtualThreadMountEnd, JVM_VirtualThreadUnmountBegin, JVM_VirtualThreadUnmountEnd, enterContinuation, yieldContinuation, looks there is no exception during mounting and unmounting, for timeout case, both virtual threads were normally unmounted, but JCL scheduler never trigger the mounting
(looks something block or interfere JCL scheduler?)
4, could not reproduce the issue with GC gencon mode and could not reproduce the issue with balanced GC with -Xint mode either, there are a couple hundred JIT optimizations has been disbabled on balanced GC due to arraylet, 
@0xdaryl  is there any JIT options which we can disable the similar optimizations in gencon mode, then we can try to isolate the issues?   Thanks

**Comment by tajila** [comment]
@VermaSh ^^ FYI

**Comment by r30shah** [comment]
Was this reproducible with the recent builds? @dchopra001  / @VermaSh  fixed an issue on Linux on Z which was causing failure in the comment https://github.com/eclipse-openj9/openj9/issues/17255#issuecomment-1526070774.

**Comment by VermaSh** [comment]
The failures seems to be because of Unsafe API. It goes away when I run with `disableUnsafe`. There aren't too many unsafe methods in the verbose log so should be able to narrow down to the failing method soon.

java cmd: 
```
$failing_build/bin/java --enable-preview -Xnocompressedrefs \
-XX:-JITServerTechPreviewMessage -Xgcpolicy:balanced -Xnoaot \
-Xjit:count=0,disableasynccompilation,verbose,vlog=pingpong.vlog,disableUnsafe \
PingPong
```

**Comment by r30shah** [comment]
@VermaSh  with respect to last comment `disableUnsafe` would inhibit bunch of optimization that would inline unsafe natives [1], not sure if you can find anything in verbose log, would recommend you to check the code base and see which unsafe transformations are not guarded by arraylets use check. 


[1]. https://github.com/search?q=repo%3Aeclipse-openj9%2Fopenj9%20TR_DisableUnsafe&type=code

**Comment by VermaSh** [comment]
The failure was actually because of `java/util/concurrent/ForkJoinPool\$WorkQueue.push(Ljava/util/concurrent/ForkJoinTask;Ljava/util/concurrent/ForkJoinPool;Z)V` and not because of Unsasfe API. I am looking into the trace log of the failing method for further investigation.

**Comment by babsingh** [comment]
> The failure was actually because of java/util/concurrent/ForkJoinPool\$WorkQueue.push(Ljava/util/concurrent/ForkJoinTask;Ljava/util/concurrent/ForkJoinPool;Z)V

@VermaSh Can you provide details on how `ForkJoinPool\$WorkQueue.push` is causing the failure?

**Comment by r30shah** [comment]
Uploading the log file here for the failing method, So far we have found that limiting compilation to that method only with even noOpt opt level is sufficient for test to get fail but when changing GC Policy with same limited method compilation works. As method has array accesses (Creates a new array for task if `ForkJoinTask` array is full and we need to insert new element, else simply inserts entry to the array in both cases it has spine checks). Trying to understand how this could have caused the failure as looking at the javacores couple of seconds apart in the hanging run, we do not see this method in the stack, most worker threads are stuck at `Unsafe.park` method. 

[ForkJoinPool.push.methodLog.txt](https://github.com/eclipse-openj9/openj9/files/11592836/ForkJoinPool.push.methodLog.txt)

**Comment by r30shah** [comment]
Babneet, Jack, Shubham an I had offline discussion about next steps in investigating this issue. So far as mentioned in the comments above, we have found out the failing method `WorkQueue.push` that pushes task to the array and signals  [1]. Method does accesses the array with BOUND and SpineCHK but looking at the core-dumps most of the arrays are 64 in size and we did not see any signs of arraylets. Shubham and I are dividing tasks here where I will try to modify the failing method and print out the worker queue to understand if it makes sense or not and Shubham will verify the generated code to make sure that we update the array in the JIT compiled code. Will keep posting updates here. 

[1]. https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L1126

**Comment by r30shah** [comment]
Couple of more things verified that Z codegen would do in compiled code. 

1. Disable inline array allocation  (Call JIT helper to allocate new array everytime)
2. Disable Arraylets (Compiled code for Balanced and Gencon would be almost same)
3. Disable inline test for ArrayStoreCHK 

With 3 above changes the test still fails. Now the only difference in generated code between gencon and balanced GC policy is the code that is generated for `awrtbari` [1] which I can not disable. Though looking at it, it is generating right code. There is one  more difference now with balanced and gencon mode is the object header size where in balanced GC policy we have data address field in the array header which we do not have in gencon. 

[1]. https://github.com/eclipse-openj9/openj9/blob/3e340dbee3b2b7d25162373e3bf88f5b41437f65/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L3616-L3689

**Comment by r30shah** [comment]
Today we did more debugging with Babneet and Jack to get to the bottom of this issue. We could still not understand how the the potential incorrect compilation of `WorkQueue.push` could cause the worker threads to be in parked mode. Adding bunch of `System.out.println` both in the JCL code as well as Unit test was causing other issues where now worker threads would be waiting on IO, misleading me to think that it failed with other other GC policy as well as Xint as well. 

We ended up deciding taking all the work JIT does in that single compiled method with which the unit test fails and see if we still get the test to fail. I also gave in-depth look at the method compilation logs collected with gencon which passes and with balanced which failed. Pasting the snippet for the ArrayStoreCHK evaluator here to explain what I found and discuss potential fix. 

```
 n186n    (  0)  ArrayStoreCHK [#43] ()                                                               [     0x3fb6f0829e0] bci=[-1,138,1146] rc=0 vc=36 vn=- li=13 udi=- nc=1 flg=0x20
 n185n    (  0)    awrtbari  <array-shadow>[#257  Shadow] [flags 0x80000607 0x0 ] ()                  [     0x3fb6f082990] bci=[-1,138,1146] rc=0 vc=36 vn=- li=13 udi=- nc=3 flg=0x20
 n184n    (  0)      aladd (X>=0 internalPtr )                                                        [     0x3fb6f082940] bci=[-1,138,1146] rc=0 vc=36 vn=- li=13 udi=- nc=2 flg=0x8100
 n165n    (  0)        ==>aload (in &GPR_0231) (X!=0 )
 n183n    (  0)        lsub                                                                           [     0x3fb6f0828f0] bci=[-1,138,1146] rc=0 vc=36 vn=- li=13 udi=- nc=2
 n181n    (  0)          ==>lshl (in GPR_0235)
 n182n    (  0)          lconst -24 (X!=0 X<=0 )                                                      [     0x3fb6f0828a0] bci=[-1,138,1146] rc=0 vc=36 vn=- li=13 udi=- nc=0 flg=0x204
 n174n    (  0)      aload  task<parm 1 Ljava/util/concurrent/ForkJoinTask;>[#397  Parm] [flags 0xc0000107 0x0 ] (in &GPR_0241)  [     0x3fb6f082620] bci=[-1,137,1146] rc=0 vc=36 vn=- li=13 udi=- nc=0
 n165n    (  0)      ==>aload (in &GPR_0231) (X!=0 )
------------------------------

 [     0x3fb6f16d2b0]	                       LGR     &GPR_0240,&GPR_0231	 ; LR=Clobber_eval
 [     0x3fb6f16d510]	                       LTG     &GPR_0241, Parm[task<parm 1 Ljava/util/concurrent/ForkJoinTask;>] ?+8(GPR5)
 [     0x3fb6f173740]	                       Label L0052:	# (Start of internal control flow)	
 [     0x3fb6f16d840]	                       CGIJ    &GPR_0241,Label L0055,0,BH(mask=0x8), 	
 [     0x3fb6f16da50]	                       BRC     NOP(0xf), Outlined Label L0056	
 [     0x3fb6f16dbf0]	                       LZRG    GPR_0238,#429 0(&GPR_0240)
 [     0x3fb6f16dd80]	                       LZRG    GPR_0239,#430 0(&GPR_0241)
 [     0x3fb6f16df10]	                       LG      GPR_0236,#431 88(GPR_0238)
 [     0x3fb6f16dfe0]	                       CLGRJ   GPR_0236,GPR_0239,Label L0053,BH(mask=0x8), 	 ; Check if src.type == array.type
 [     0x3fb6f16e240]	                       CLGRL   GPR_0236, 0x000003FB6F16E0F0	
 [     0x3fb6f16e350]	                       BRC     BH(0x8), Label L0053	 ; Check if array.type is type object, if yes jump to wbLabel
 [     0x3fb6f16e510]	                       CLG     GPR_0236,#432 200(GPR_0239)
 [     0x3fb6f16e5e0]	                       BRC     BH(0x8), Label L0053	 ; Check if src.class(in t1Reg).castClassCache == array.componentClass
 [     0x3fb6f16e7a0]	                       LLGH    GPR_0238,#433 30(GPR_0236)
 [     0x3fb6f16e930]	                       LLGH    GPR_0237,#434 30(GPR_0239) ; Check if src.type is subclass Failure if depth(src) < depth(array-type)
 [     0x3fb6f16ea00]	                       CGRJ    GPR_0238,GPR_0237,Outlined Label L0056,BL(mask=0x2), 	
 [     0x3fb6f16eb10]	                       SLLG    GPR_0238,GPR_0238,3
 [     0x3fb6f16ecb0]	                       LG      GPR_0237,#435 16(GPR_0239)
 [     0x3fb6f16ee40]	                       CLG     GPR_0236,#436 0(GPR_0238,GPR_0237)
 [     0x3fb6f16ef20]	                       BRC     BRE(0x7), Outlined Label L0056	
 [     0x3fb6f1724c0]	                       Label L0053:	
 [     0x3fb6f1725a0]	                       STG     &GPR_0241, Shadow[<array-shadow>] 24(GPR_0235,&GPR_0231)
 [     0x3fb6f172730]	                       SG      &GPR_0240,#437 216(GPR13)
 [     0x3fb6f1728c0]	                       CLG     &GPR_0240,#438 224(GPR13)
 [     0x3fb6f172990]	                       BRC     BNM(0xa), Label L0054	
 [     0x3fb6f172a70]	                       SRLG    &GPR_0240,&GPR_0240,9
 [     0x3fb6f172c10]	                       AG      &GPR_0240,#439 200(GPR13)
 [     0x3fb6f172da0]	                       MVI     #440 0(&GPR_0240), 0x01
 [     0x3fb6f172e80]	                       Label L0055:	
 [     0x3fb6f172fe0]	                       STG     &GPR_0241, Shadow[<array-shadow>] 24(GPR_0235,&GPR_0231)
 [     0x3fb6f173670]	                       assocreg
 [     0x3fb6f1730b0]	                       Label L0054:	# (End of internal control flow)	
 POST:
 {GPR1:&GPR_0240:R} {GPR2:&GPR_0241:R} {GPR14:GPR_0236:R} {GPR4:GPR_0237:R} {AssignAny:GPR_0239:R} {AssignAny:GPR_0238:R} {AssignAny:&GPR_0231:R} {AssignAny:GPR_0235:R}

------------------------------

 [     0x3fb6f16f050]	                       Outlined Label L0056:	
 [     0x3fb6f170cd0]	                       LGR     GPR_0282,GPR13	
 [     0x3fb6f170e60]	                       LGR     &GPR_0283,&GPR_0231	 ; LR=Clobber_eval
 [     0x3fb6f171930]	                       assocreg
 [     0x3fb6f172310]	                       assocreg
  PRE:
 {GPR2:GPR_0282:R} {GPR3:&GPR_0241:R} {GPR4:&GPR_0283:R}
 [     0x3fb6f171340]	                       BRASL   GPR_0249,0x0000000000000000	
 POST:
 {GPR0:D_GPR_0244:D}* {GPR1:D_GPR_0245:D}* {GPR2:D_GPR_0246:D}* {GPR3:D_GPR_0247:D}* {GPR4:D_GPR_0248:D}* {GPR14:GPR_0249:D} {FPR0:D_FPR_0250:D}* {FPR1:D_FPR_0251:D}*
 {FPR2:D_FPR_0252:D}* {FPR3:D_FPR_0253:D}* {FPR4:D_FPR_0254:D}* {FPR5:D_FPR_0255:D}* {FPR6:D_FPR_0256:D}* {FPR7:D_FPR_0257:D}* {FPR8:D_FPR_0258:D}* {FPR9:D_FPR_0259:D}*
 {FPR10:D_FPR_0260:D}* {FPR11:D_FPR_0261:D}* {FPR12:D_FPR_0262:D}* {FPR13:D_FPR_0263:D}* {FPR14:D_FPR_0264:D}* {FPR15:D_FPR_0265:D}* {VRF16:D_VRF_0266:D}* {VRF17:D_VRF_0267:D}*
 {VRF18:D_VRF_0268:D}* {VRF19:D_VRF_0269:D}* {VRF20:D_VRF_0270:D}* {VRF21:D_VRF_0271:D}* {VRF22:D_VRF_0272:D}* {VRF23:D_VRF_0273:D}* {VRF24:D_VRF_0274:D}* {VRF25:D_VRF_0275:D}*
 {VRF26:D_VRF_0276:D}* {VRF27:D_VRF_0277:D}* {VRF28:D_VRF_0278:D}* {VRF29:D_VRF_0279:D}* {VRF30:D_VRF_0280:D}* {VRF31:D_VRF_0281:D}*
 [     0x3fb6f1723e0]	                       BRC     NOP(0xf), Label L0053	
```


In above sequence write barrier code we generate from `awrtbari` starts from `Label L0053:`. Looking at the markdown document from the JIT write barrier [1], code seems to be doing the write thing, but once interesting thing I saw which we do not do in case gc policy is `gencon` is re performing store to array at instruction `0x3fb6f172fe0` in balanced. Looking into the reasoning behind that (Luckily a code does have a comment behind the purpose) [2], it seems like that bypass branch was omitted in case of balanced as inserting to branching around store to memory might be cheaper. I do not we do store to that object twice atleast on x86 so it definitely looks like we tried to optimize the generated code to make is fall through to perform second store which was intended to be executed when we are storing NULL in the array for which we can skip the write barrier. 

This commit https://github.com/r30shah/openj9/commit/1820134f214e6ea106a90daa563e963a09d4fe2a contains the potential fix which I have verified in my environment to fix the issue (so far passes 20 runs out of 20 runs). Will launch other set of sanity builds tomorrow morning to put it to further tests before opening up PR. 


[1]. https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/concepts/JitWriteBarriers.md
[2]. https://github.com/eclipse-openj9/openj9/blob/bedafef8715df321eb2ee7fd5c64af7471fb3861/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L7184-L7198

**Comment by joransiu** [comment]
Interesting... The only reason I can think of is that another thread does a store or some atomic update operation in between the two stores in this array write barrier sequence.  With that second store, we've effectively restored the old "stale" value incorrectly into the array.

**Comment by r30shah** [comment]
I was thinking same, so tried to correlate how current work could have triggered the incorrect condition. Analyzing the locations in the `ForkJoinPool.java` where it could have call the `queue.push` method [1][2]. It calls the method `submissionQueue` [3] which will returns a queue to which a task will be added to. This method will take care of either creating a new work queue object with array of ForkJoinTask to add task to if we do not have one or checkout the available queue from the array which will  hold the task. Only case where this method will return with correct WorkQueue object seems to be at [4] where it will set the access flag in the Queue to 1 and if it can not do it, it will pick up another queue that is available. Checking out the code path for `getAndSetAccess(1)`, it would atomically update the `access` field using Unsafe that seems correct to me. 

One of the location where the task would be worked on is in [WorkQueue.scan](https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L1837-L1845) where it will atomically set the task to NULL after reading it and performs it. It may be possible when we have first store the task in the array and then performing the code for write barrier, it will read the task and sets it to NULL, which will later gets overwritten when we do second time store in the write barrier, causing unexpected state.  I am not sure if this is what is happening, but I think re-writing into same array field again would be anyway seems to be not required especially for the purpose of write barrier, and I should continue with the PR to fix this. 

[1]. https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2358-L2372
[2]. https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2922-L2928
[3]. https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2922-L2928
[4]. https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2341

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/issues/17255#issuecomment-1573118328 and https://github.com/eclipse-openj9/openj9/issues/17255#issuecomment-1574193482:

[Minor correction] [WorkQueue.scan](https://github.com/ibmruntimes/openj9-openjdk-jdk20/blob/a80d1107c64a663a83e63bf60cac065b29d93d81/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L1837-L1845): `scan` is a member of the `ForkJoinPool` class. Every `ForkJoinWorkerThread` invokes `ForkJoinPool.scan` to look for work i.e. scan every `WorkQueue`. The CAS in `ForkJoinPool.scan` is the source of the atomic operation, where the `WorkQueue.array` element is set to NULL after a `ForkJoinTask` (work) is found. This CAS can happen between the two stores in `WorkQueue.push's` array write barrier sequence.

**Comment by pshipton** [comment]
Re-open until the fix is delivered to 0.39.
